packages/medusa/src/services/shipping-option.ts

```diff
@@ -0,0 +1,717 @@
+      option.amount =
+        data.price_type === "calculated" ? null : data.amount ?? null
+
+      const isValid = await this.providerService_.validateOption(option)
+
+      if (!isValid) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "The fulfillment provider cannot validate the shipping option"
+        )
+      }
+
+      if (typeof data.requirements !== "undefined") {
+        const acc: ShippingOptionRequirement[] = []
+        for (const r of data.requirements) {
+          const validated = await this.validateRequirement_(r)
+
+          if (acc.find((raw) => raw.type === validated.type)) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              "Only one requirement of each type is allowed"
+            )
+          }
+
+          if (
+            acc.find(
+              (raw) =>
+                (raw.type === "max_subtotal" &&
+                  validated.amount > raw.amount) ||
+                (raw.type === "min_subtotal" && validated.amount < raw.amount)
+            )
+          ) {
+            throw new MedusaError(
+              MedusaError.Types.INVALID_DATA,
+              "Max. subtotal must be greater than Min. subtotal"
+            )
+          }
+
+          acc.push(validated)
+        }
+      }
+
+      const result = await optionRepo.save(option)
+      return result
+    })
+  }
+
+  /**
+   * Validates a shipping option price
+   * @param {ShippingOptionPriceType} priceType - the price to validate
+   * @param {ShippingOption} option - the option to validate against
+   * @return {Promise<ShippingOptionPriceType>} the validated price
+   */
+  async validatePriceType_(
+    priceType: ShippingOptionPriceType,
+    option: ShippingOption
+  ): Promise<ShippingOptionPriceType> {
+    if (
+      !priceType ||
+      (priceType !== "flat_rate" && priceType !== "calculated")
+    ) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "The price must be of type flat_rate or calculated"
+      )
+    }
+
+    if (priceType === "calculated") {
+      const canCalculate = await this.providerService_.canCalculate(option)
+      if (!canCalculate) {
+        throw new MedusaError(
+          MedusaError.Types.INVALID_DATA,
+          "The fulfillment provider cannot calculate prices for this option"
+        )
+      }
+    }
+
+    return priceType
+  }
+
+  /**
+   * Updates a profile. Metadata updates and product updates should use
+   * dedicated methods, e.g. `setMetadata`, etc. The function
+   * will throw errors if metadata or product updates are attempted.
+   * @param {string} optionId - the id of the option. Must be a string that
+   *   can be casted to an ObjectId
+   * @param {object} update - an object with the update values.
+   * @return {Promise} resolves to the update result.
+   */
+  async update(
+    optionId: string,
+    update: UpdateShippingOptionProps
```

Comment:
very nit(non-blocking): Naming
```suggestion
    update: UpdateShippingOptionInput
```